feat: Add Azure MongoDBCosmosDB & ChromaDB Integrations

ishaansehgal99 opened the pull request:
### **User description**
**Reason for Change**:
With new base class introduces a new popular integration - chromadb and azurecosmosdb mongodb

<i>PR also adds license headers</i>


___

### **PR Type**
Enhancement, Tests


___

### **Description**
- Added tests for Azure MongoDB and ChromaDB integrations.

- Implemented new vector store handlers for Azure MongoDB and ChromaDB.

- Updated ChromaDB store to support asynchronous operations.

- Added license headers to relevant files.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_azuremongodb_store.py</strong><dd><code>Add Azure MongoDB Integration Tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py

<li>Added tests for Azure MongoDB vector store.<br> <li> Implemented test cases for indexing, querying, and document existence.


</details>


  </td>
  <td><a href="https://github.com/kaito-project/kaito/pull/649/files#diff-5a7d74c6456562a8ccaf3c676ae087e63956bc5bb7ffa3a3f42b8cfc25a9b715">+126/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>azuremongodb_store.py</strong><dd><code>Implement Azure MongoDB Vector Store Handler</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pkg/ragengine/services/vector_store/azuremongodb_store.py

<li>Implemented Azure MongoDB vector store handler.<br> <li> Added methods for creating indices, listing documents, and checking <br>document existence.


</details>


  </td>
  <td><a href="https://github.com/kaito-project/kaito/pull/649/files#diff-51b0df0709ce0ae54452657c87e57cd1a80e39821466db2b9563496debf288a5">+93/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_chromadb_store.py</strong><dd><code>Update ChromaDB Store for Async Operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

presets/ragengine/tests/vector_store/test_chromadb_store.py

<li>Updated ChromaDB store to support asynchronous operations.<br> <li> Added asynchronous methods for listing documents.


</details>


  </td>
  <td><a href="https://github.com/kaito-project/kaito/pull/649/files#diff-b2994135bd54f8de9acf5019a3555abb8e9439a7acb606633185610410f062b7">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>chromadb_store.py</strong><dd><code>Update ChromaDB Store for Async Operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

presets/ragengine/vector_store/chromadb_store.py

<li>Updated ChromaDB store to support asynchronous operations.<br> <li> Added asynchronous methods for listing documents.


</details>


  </td>
  <td><a href="https://github.com/kaito-project/kaito/pull/649/files#diff-dc5b977536114136265949c77411399249a2cde5417a57ffeaabfce46644638b">+30/-8</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

zhuangqh commented on ragengine/vector_store/base.py:
```diff
@@ -1,31 +1,143 @@
+            index.set_index_id(index_name)
+            self.index_map[index_name] = index
+            self.index_store.add_index_struct(index.index_struct)
+            self._persist(index_name)
```
I am thinking about a question: will the persist fails here? or we should do some rollback.

ishaansehgal99 replied:
its possible, added a try-catch in persist function to help prevent this

zhuangqh replied:
if persist fails, should we rollback the index_store. Or there will be some inconsistencies between index_store and storage

ishaansehgal99 replied:
storage is just for debugging purposes, there shouldn't be inconsistencies

zhuangqh commented on ragengine/vector_store/base.py:
```diff
@@ -1,31 +1,143 @@
+                self.add_document_to_index(index_name, doc, doc_id)
+                indexed_doc_ids.add(doc_id)
+            else:
+                print(f"Document {doc_id} already exists in index {index_name}. Skipping.")
```
nit: add a logger to control these messages

ishaansehgal99 replied:
done

ishaansehgal99 commented:
Preparing review...

ishaansehgal99 commented:
## PR Reviewer Guide 🔍

Here are some key observations to aid the review process:

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 4 🔵🔵🔵🔵⚪</td></tr>
<tr><td>🧪&nbsp;<strong>PR contains tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>Security concerns</strong><br><br>

<strong>Sensitive information exposure:</strong><br> The code uses a hardcoded MongoDB URI in the environment variable `AZURE_COSMOSDB_MONGODB_URI`, which could expose sensitive information if the environment is not properly secured.</td></tr>
<tr><td>⚡&nbsp;<strong>Recommended focus areas for review</strong><br><br>

<details><summary><a href='https://github.com/kaito-project/kaito/pull/649/files#diff-51b0df0709ce0ae54452657c87e57cd1a80e39821466db2b9563496debf288a5R26-R26'><strong>Potential Security Issue</strong></a>

The code uses a hardcoded MongoDB URI in the environment variable `AZURE_COSMOSDB_MONGODB_URI`. This could expose sensitive information if the environment is not properly secured.
</summary>

```python
self.connection_string = os.getenv("AZURE_COSMOSDB_MONGODB_URI", "mongodb+srv://myDatabaseUser:D1fficultP%40ssw0rd@cluster0.example.mongodb.net/?retryWrites=true&w=majority")
```

</details>

<details><summary><a href='https://github.com/kaito-project/kaito/pull/649/files#diff-51b0df0709ce0ae54452657c87e57cd1a80e39821466db2b9563496debf288a5R29-R31'><strong>Missing Error Handling</strong></a>

The code does not handle exceptions that could occur when interacting with the MongoDB client, such as network issues or invalid credentials.
</summary>

```python
    self.mongodb_client = pymongo.MongoClient(self.connection_string)
    self.mongodb_client.admin.command('ping') # Test the connection
except Exception as e:
```

</details>

</td></tr>
</table>

ishaansehgal99 commented:
Preparing answer...

ishaansehgal99 commented:
### **Ask**❓
what is this pr doing

### **Answer:**
This PR is adding integrations for Azure MongoDB CosmosDB and ChromaDB to the existing vector store service. The main changes include:

1. **New Test Files**: Two new test files (`test_azuremongodb_store.py` and `test_chromadb_store.py`) have been added to test the new integrations.

2. **New Vector Store Handler Classes**: Two new classes (`AzureCosmosDBMongoDBVectorStoreHandler` and `ChromaDBVectorStoreHandler`) have been added to handle the vector store operations for Azure MongoDB CosmosDB and ChromaDB, respectively.

3. **License Headers**: License headers have been added to all the new files.

### Detailed Breakdown:

#### `test_azuremongodb_store.py`
- **Test Fixtures**: The file includes fixtures to initialize the embedding manager and the vector store manager.
- **Test Functions**: Several test functions are defined to test various operations like indexing documents, querying documents, and checking document existence.

#### `azuremongodb_store.py`
- **Class Definition**: A new class `AzureCosmosDBMongoDBVectorStoreHandler` is defined, inheriting from `BaseVectorStore`.
- **Initialization**: The class initializes the MongoDB client and ensures the collection exists.
- **Indexing**: The `_create_new_index` method creates a new index in MongoDB.
- **Listing Documents**: The `list_all_indexed_documents` method lists all indexed documents.
- **Document Existence**: The `document_exists` method checks if a document exists in the index.
- **Cleanup**: The `_clear_collection_and_indexes` method clears all documents and drops all indexes in the collection.

#### `test_chromadb_store.py`
- **Test Fixtures**: Similar to the Azure MongoDB test file, this file includes fixtures to initialize the embedding manager and the vector store manager.
- **Test Functions**: The `check_indexed_documents` function is defined to test listing all indexed documents.

#### `chromadb_store.py`
- **Class Definition**: A new class `ChromaDBVectorStoreHandler` is defined, inheriting from `BaseVectorStore`.
- **Initialization**: The class initializes the ChromaDB client.
- **Indexing**: The `_create_new_index` method creates a new index in ChromaDB.
- **Listing Documents**: The `list_documents_in_index` and `list_all_documents` methods list documents in a specific index and all indexes, respectively.
- **Document Existence**: The `document_exists` method checks if a document exists in the index.
- **Cleanup**: The `_clear_collection_and_indexes` method clears all collections in the ChromaDB instance.

### Summary
The PR adds support for two new vector store integrations, Azure MongoDB CosmosDB and ChromaDB, by introducing new classes, test files, and ensuring proper initialization and cleanup of the vector store operations.

ishaansehgal99 commented:
## Generating PR code suggestions


Work in progress ...<br>
<img src="https://codium.ai/images/pr_agent/dual_ball_loading-crop.gif" width=48>

ishaansehgal99 commented:
Generating Documentation...

ishaansehgal99 commented on pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py:
```diff
@@ -0,0 +1,126 @@
+
+from services.vector_store.base import BaseVectorStore
+from services.vector_store.azuremongodb_store import AzureCosmosDBMongoDBVectorStoreHandler
+from services.models import Document
```
**Suggestion:** Proposed documentation
```suggestion
from services.models import Document
"""
Test suite for the AzureCosmosDBMongoDBVectorStoreHandler class.
"""
```

ishaansehgal99 commented on pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py:
```diff
@@ -0,0 +1,126 @@
+
+@pytest.fixture
+def vector_store_manager(init_embed_manager):
+    with TemporaryDirectory() as temp_dir:
```
**Suggestion:** Proposed documentation
```suggestion
    with TemporaryDirectory() as temp_dir:
        """
        Fixture to initialize the embedding manager for vector store tests.
        """
```

ishaansehgal99 commented on pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py:
```diff
@@ -0,0 +1,126 @@
+        manager = AzureCosmosDBMongoDBVectorStoreHandler(init_embed_manager)
+        manager._clear_collection_and_indexes()
+        yield manager
+
```
**Suggestion:** Proposed documentation
```suggestion
"""
Fixture to initialize the vector store manager for vector store tests.
"""
```

ishaansehgal99 commented on pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py:
```diff
@@ -0,0 +1,126 @@
+        Document(text=first_doc_text, metadata={"type": "text"}),
+        Document(text=second_doc_text, metadata={"type": "text"})
+    ]
+    
```
**Suggestion:** Proposed documentation
```suggestion
    
    """
    Tests indexing documents into the vector store.
    """
```

ishaansehgal99 commented on pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py:
```diff
@@ -0,0 +1,126 @@
+    assert set(doc_ids) == {BaseVectorStore.generate_doc_id(first_doc_text),
+                            BaseVectorStore.generate_doc_id(second_doc_text)}
+
+def test_index_documents_isolation(vector_store_manager):
```
**Suggestion:** Proposed documentation
```suggestion
def test_index_documents_isolation(vector_store_manager):
    """
    Tests document isolation across different indices.
    """
```

ishaansehgal99 commented on pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py:
```diff
@@ -0,0 +1,126 @@
+    vector_store_manager.index_documents(index_name_1, documents1)
+    vector_store_manager.index_documents(index_name_2, documents2)
+
+    indexed_docs = vector_store_manager.list_all_indexed_documents()
```
**Suggestion:** Proposed documentation
```suggestion
    indexed_docs = vector_store_manager.list_all_indexed_documents()
    """
    Tests querying documents using the vector store.
    """
```

ishaansehgal99 commented on pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py:
```diff
@@ -0,0 +1,126 @@
+    mock_response = {
+        "result": "This is the completion from the API"
+    }
+
```
**Suggestion:** Proposed documentation
```suggestion
"""
Tests adding a document to an existing index.
"""
```

ishaansehgal99 commented on pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py:
```diff
@@ -0,0 +1,126 @@
+    documents = [
+        Document(text="First document", metadata={"type": "text"}),
+        Document(text="Second document", metadata={"type": "text"})
+    ]
```
**Suggestion:** Proposed documentation
```suggestion
    ]
    """
    Tests persisting a single index.
    """
```

ishaansehgal99 commented on pkg/ragengine/services/tests/vector_store/test_azuremongodb_store.py:
```diff
@@ -0,0 +1,126 @@
+    # Mock query and results
+    query_result = vector_store_manager.query("test_index", "First", top_k=1, llm_params=params)
+
+    assert query_result is not None
```
**Suggestion:** Proposed documentation
```suggestion
    assert query_result is not None
    """
    Tests persisting multiple indices.
    """
```

ishaansehgal99 commented on pkg/ragengine/services/vector_store/azuremongodb_store.py:
```diff
@@ -0,0 +1,93 @@
+logger = logging.getLogger(__name__)
+
+class AzureCosmosDBMongoDBVectorStoreHandler(BaseVectorStore):
+    def __init__(self, embedding_manager):
```
**Suggestion:** Proposed documentation
```suggestion
    def __init__(self, embedding_manager):
        """
        Handler class for interacting with Azure Cosmos DB MongoDB for vector storage.
        """
```

ishaansehgal99 commented on pkg/ragengine/services/vector_store/azuremongodb_store.py:
```diff
@@ -0,0 +1,93 @@
+        except Exception as e:
+            raise ValueError(f"Failed to access collection '{self.collection_name}' in database '{self.db_name}': {e}")
+
+
```
**Suggestion:** Proposed documentation
```suggestion
"""
Initializes the Azure Cosmos DB MongoDB vector store handler.
"""
```

ishaansehgal99 commented on pkg/ragengine/services/vector_store/azuremongodb_store.py:
```diff
@@ -0,0 +1,93 @@
+            db_name=self.db_name,
+            collection_name=self.collection_name,
+            index_name=index_name,
+            embedding_key=f"{index_name}_embedding",  # Unique field for each index
```
**Suggestion:** Proposed documentation
```suggestion
            embedding_key=f"{index_name}_embedding",  # Unique field for each index
            """
            Creates a new index in the vector store.
            """
```

ishaansehgal99 commented on pkg/ragengine/services/vector_store/azuremongodb_store.py:
```diff
@@ -0,0 +1,93 @@
+        return self._create_index_common(index_name, documents, vector_store)
+
+    def list_all_indexed_documents(self) -> Dict[str, Dict[str, Dict[str, str]]]:
+        indexed_docs = {}  # Accumulate documents across all indexes
```
**Suggestion:** Proposed documentation
```suggestion
        indexed_docs = {}  # Accumulate documents across all indexes
        """
        Lists all indexed documents in the vector store.
        """
```

ishaansehgal99 commented on pkg/ragengine/services/vector_store/azuremongodb_store.py:
```diff
@@ -0,0 +1,93 @@
+                if doc_id is None:
+                    continue  # Skip if no document ID is found
+                indexed_docs.setdefault(index_name, {})[doc_id] = {
+                    "text": doc.get("text", ""),
```
**Suggestion:** Proposed documentation
```suggestion
                    "text": doc.get("text", ""),
                    """
                    Checks if a document exists in the vector store.
                    """
```

ishaansehgal99 commented on pkg/ragengine/services/vector_store/azuremongodb_store.py:
```diff
@@ -0,0 +1,93 @@
+        return indexed_docs
+
+    def document_exists(self, index_name: str, doc: Document, doc_id: str) -> bool:
+        """AzureCosmosDBMongoDB for checking document existence."""
```
**Suggestion:** Proposed documentation
```suggestion
        """AzureCosmosDBMongoDB for checking document existence."""
        """
        Clears all documents and drops all indexes in the vector store.
        """
```

ishaansehgal99 commented on presets/ragengine/tests/vector_store/test_chromadb_store.py:
```diff
@@ -21,8 +21,9 @@ def vector_store_manager(self, init_embed_manager):
 
-    def check_indexed_documents(self, vector_store_manager):
-        indexed_docs = vector_store_manager.list_all_indexed_documents()
+    @pytest.mark.asyncio
```
**Suggestion:** Proposed documentation
```suggestion
    @pytest.mark.asyncio
    """
    Tests listing all indexed documents in the vector store.
    """
```

ishaansehgal99 commented on presets/ragengine/vector_store/chromadb_store.py:
```diff
@@ -20,23 +21,45 @@ def __init__(self, embedding_manager):
+        collection = await asyncio.to_thread(
+            lambda: self.chroma_client.get_collection(index_name).get()
+        )
+        return doc.text in collection["documents"]
```
**Suggestion:** Proposed documentation
```suggestion
        return doc.text in collection["documents"]
"""
Creates a new index in the vector store.
"""
```

ishaansehgal99 commented on presets/ragengine/vector_store/chromadb_store.py:
```diff
@@ -20,23 +21,45 @@ def __init__(self, embedding_manager):
+                doc_map[doc[0]] = {
+                    "text": doc[1],
+                    "metadata": json.dumps(doc[2])
+                }
```
**Suggestion:** Proposed documentation
```suggestion
                }
        """
        Checks if a document exists in the vector store.
        """
```

ishaansehgal99 commented on presets/ragengine/vector_store/chromadb_store.py:
```diff
@@ -20,23 +21,45 @@ def __init__(self, embedding_manager):
+            collection_names = await asyncio.to_thread(self.chroma_client.list_collections)
+            for collection_name in collection_names:
+                collection_info = await asyncio.to_thread(
+                    lambda: self.chroma_client.get_collection(collection_name).get()
```
**Suggestion:** Proposed documentation
```suggestion
                    lambda: self.chroma_client.get_collection(collection_name).get()
                """
                Lists all documents in a specific index.
                """
```

ishaansehgal99 commented:
Preparing PR description...

ishaansehgal99 commented:
Preparing PR description...

ishaansehgal99 commented:
Preparing PR labels...

ishaansehgal99 commented:
Preparing PR labels...

ishaansehgal99 commented:
Preparing PR description...

ishaansehgal99 commented:
**[PR Description](https://github.com/kaito-project/kaito/pull/649)** updated to latest commit (https://github.com/kaito-project/kaito/commit/d60dba23b86dd63ee35c7a82dfd93e2699abda65)

ishaansehgal99 commented:
Preparing changelog updates...

ishaansehgal99 commented:
**Changelog updates:** 🔄

## 2025-03-05

### Added
- Added tests for Azure MongoDB and ChromaDB integrations.
- Implemented new vector store handlers for Azure MongoDB and ChromaDB.
- Updated ChromaDB store to support asynchronous operations.
- Added license headers to relevant files.


>to commit the new content to the CHANGELOG.md file, please type:
>'/update_changelog --pr_update_changelog.push_changelog_changes=true'

ishaansehgal99 commented:
Preparing answer...

ishaansehgal99 commented:
### **Ask**❓
write me a poem about this PR

### **Answer:**
### A Poem About the PR

In a world of code and data,
A new integration was added, a grand plan.
Azure MongoDB and ChromaDB,
Now part of our vector store, a new chapter.

Tests were added, a suite so fine,
Ensuring our new features shine.
Indexing, querying, and more,
Our vector store now has a new core.

ChromaDB's async support,
A step forward, a leap in progress.
With ChromaDB's async support,
Our vector store now has a new boost.

License headers, a must,
Ensuring our code is open and trust.
With license headers in place,
Our code is now free to race.

A new feature, a new addition,
To our vector store, a new direction.
Azure MongoDB and ChromaDB,
Now part of our vector store, a new chapter.

ishaansehgal99 commented:
Preparing review...

ishaansehgal99 commented:
**[Persistent review](https://github.com/kaito-project/kaito/pull/649#issuecomment-2702452896)** updated to latest commit https://github.com/kaito-project/kaito/commit/d60dba23b86dd63ee35c7a82dfd93e2699abda65

ishaansehgal99 commented:
Outdated, will be addressed in new PR
